pkg/ebpf/c/tcp-queue-length-kern.c

```diff
@@ -0,0 +1,136 @@
+#include <linux/kconfig.h>
+#define KBUILD_MODNAME "foo"
+#include <linux/ptrace.h>
+#include <linux/bpf.h>
+#include <net/inet_sock.h>
+#include <linux/tcp.h>
+
+#include "tcp-queue-length-kern-user.h"
+
+BPF_HASH(queue, struct sock *, struct stats);
+
+BPF_HASH(who_recvmsg, u64, struct sock *);
+BPF_HASH(who_sendmsg, u64, struct sock *);
+
+// TODO: replace all `bpf_probe_read` by `bpf_probe_read_kernel` once we can assume that we have at least kernel 5.5
+static inline int check_sock(struct sock *sk) {
+  struct stats zero = {
+    .rqueue = {
+      .min = 2^32-1,
+      .max = 0
+    },
+    .wqueue = {
+      .min = 2^32-1,
+      .max = 0
+    }
+  };
+
+  struct stats *s = queue.lookup_or_init(&sk, &zero);
+  if (s == NULL) return 0;
+
+  /*
+   * We assume here that only one thread will read and/or write to a given socket.
+   * Indeed, having several unsynchronized threads attempting to read and/or write to a socket
+   * would corrupt the stream.
+   * If that assumption was wrong, we would need to make the following piece of code thread safe.
+   * In that case, per-cpu hash would be a better solution than mutex.
+   */
+  if (s->pid == 0) {
+    s->pid = bpf_get_current_pid_tgid() >> 32;
+
+    struct task_struct *cur_tsk = (struct task_struct *)bpf_get_current_task();
+    struct css_set *css_set;
+    if (!bpf_probe_read(&css_set, sizeof(css_set), &cur_tsk->cgroups)) {
+      struct cgroup_subsys_state *css;
+      // TODO: Do not arbitrarily pick the first subsystem
```

Comment:
It’s just that, from theoretical point of view, using the first subsystem sounds arbitrary and it may look more deterministic to look for a precise one. But even in that case, why would looking for the `cpu` cgroup make more sense than looking for the `memory` cgroup ?

The issue is that doing this would require to loop over the cgroup subsystems and loops are not allowed in eBPF so that the resulting code would be significantly more cumbersome.

In practice, it doesn’t matter at all since container engines are putting the container processes  in the same path in all subsystems:
```
lenaic.huard@gke-lenaic-ubuntu-docker-cb7abb6a-f89l:~$ sudo cat /proc/2364203/cgroup
11:freezer:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
10:net_cls,net_prio:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
9:blkio:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
8:perf_event:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
7:cpuset:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
6:cpu,cpuacct:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
5:memory:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
4:devices:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
3:pids:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
2:hugetlb:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
1:name=systemd:/kubepods/besteffort/pod0daa7e79-ae25-45aa-ac7e-8159814f86f3/bd093b912581e638a6c913697c2357e92fdeef0dcbf2f0cb788eae6d6dd01ef2
0::/system.slice/containerd.service
```
(The path for id 0 is the cgroup v2 hierarchy and cgroup v2 are in different fields. This piece of code doesn’t access the cgroup v2 informations)